Altair: Incorrectly requesting sync committee duties prior to fork activation

### Description
With a minimal configuration and `ALTAIR_FORK_SLOT: 8`, the chain starts correctly with no exceptions but at slot 8 we start getting exceptions because the validator client is attempting to load sync committee duties for epoch 0 which is prior to the Altair fork:

```
2021-05-07 10:09:54.016 | validator-async-7         | ERROR | RetryingDutyLoader             | Failed to request validator duties for epoch 0. Retrying after delay.
java.util.concurrent.CompletionException: java.lang.IllegalStateException: Fork at slot 0 does not support sync committees
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1155) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147) ~[classes/:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
	at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:529) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:32) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:142) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$27(SafeFuture.java:367) ~[classes/:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147) ~[classes/:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateToAsync$13(SafeFuture.java:269) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:81) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$2(AsyncRunner.java:38) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:73) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:119) ~[classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: java.lang.IllegalStateException: Fork at slot 0 does not support sync committees
	at tech.pegasys.teku.spec.Spec.lambda$getSyncCommitteeUtilRequired$1(Spec.java:143) ~[classes/:?]
	at java.util.Optional.orElseThrow(Optional.java:401) ~[?:?]
	at tech.pegasys.teku.spec.Spec.getSyncCommitteeUtilRequired(Spec.java:141) ~[classes/:?]
	at tech.pegasys.teku.validator.client.SyncCommitteeDutyLoader.scheduleAllDuties(SyncCommitteeDutyLoader.java:65) ~[classes/:?]
	at tech.pegasys.teku.validator.client.SyncCommitteeDutyLoader.scheduleAllDuties(SyncCommitteeDutyLoader.java:27) ~[classes/:?]
	at tech.pegasys.teku.validator.client.AbstractDutyLoader.lambda$loadDutiesForEpoch$2(AbstractDutyLoader.java:54) ~[classes/:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1146) ~[?:?]
	... 27 more
```

--- a/ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtil.java
+++ b/ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtil.java
@@ -80,6 +80,9 @@ public class SyncCommitteeUtil {
 
   public boolean isStateUsableForCommitteeCalculationAtEpoch(
       final BeaconState state, final UInt64 epoch) {
+    if (state.toVersionAltair().isEmpty()) {
+      return false;
+    }
     final UInt64 syncCommitteePeriod = computeSyncCommitteePeriod(epoch);
     final UInt64 currentEpoch = beaconStateAccessors.getCurrentEpoch(state);
     final UInt64 currentSyncCommitteePeriod = computeSyncCommitteePeriod(currentEpoch);

--- a/ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtilTest.java
+++ b/ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtilTest.java
@@ -30,6 +30,7 @@ import tech.pegasys.teku.spec.TestSpecFactory;
 import tech.pegasys.teku.spec.config.SpecConfigAltair;
 import tech.pegasys.teku.spec.datastructures.state.Validator;
 import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.altair.BeaconStateAltair;
 import tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.altair.BeaconStateSchemaAltair;
 import tech.pegasys.teku.spec.datastructures.type.SszPublicKey;
 import tech.pegasys.teku.spec.datastructures.util.SyncSubcommitteeAssignments;
@@ -187,10 +188,10 @@ class SyncCommitteeUtilTest {
 
   @Test
   void shouldComputeFirstEpochOfNextSyncCommitteePeriod() {
-    final int epocsPerPeriod = config.getEpochsPerSyncCommitteePeriod();
+    final int epochsPerPeriod = config.getEpochsPerSyncCommitteePeriod();
     final UInt64 period1Start = UInt64.ZERO;
-    final UInt64 period2Start = period1Start.plus(epocsPerPeriod);
-    final UInt64 period3Start = period2Start.plus(epocsPerPeriod);
+    final UInt64 period2Start = period1Start.plus(epochsPerPeriod);
+    final UInt64 period3Start = period2Start.plus(epochsPerPeriod);
 
     assertThat(syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(period1Start))
         .isEqualTo(period2Start);
@@ -203,6 +204,58 @@ class SyncCommitteeUtilTest {
         .isEqualTo(period3Start);
     assertThat(syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(period2Start.plus(1)))
         .isEqualTo(period3Start);
+  }
+
+  @Test
+  void isStateUsableForCommitteeCalculationAtEpoch_shouldBeFalseWhenStateIsNotFromAltairFork() {
+    assertThat(
+            syncCommitteeUtil.isStateUsableForCommitteeCalculationAtEpoch(
+                dataStructureUtil.randomBeaconState(UInt64.ZERO), UInt64.ZERO))
+        .isFalse();
+  }
+
+  @Test
+  void isStateUsableForCommitteeCalculationAtEpoch_shouldBeFalseWhenStateFromNextPeriod() {
+    final int epochsPerPeriod = config.getEpochsPerSyncCommitteePeriod();
+    final UInt64 period1Start = UInt64.ZERO;
+    final UInt64 period2Start = spec.computeStartSlotAtEpoch(period1Start.plus(epochsPerPeriod));
+
+    final BeaconStateAltair state =
+        dataStructureUtil.stateBuilderAltair().slot(period2Start).build();
+    assertThat(syncCommitteeUtil.isStateUsableForCommitteeCalculationAtEpoch(state, UInt64.ZERO))
+        .isFalse();
+  }
+
+  @Test
+  void isStateUsableForCommitteeCalculationAtEpoch_shouldBeFalseWhenStateFromTwoPeriodsAgo() {
+    final int epochsPerPeriod = config.getEpochsPerSyncCommitteePeriod();
+
+    final BeaconStateAltair state =
+        dataStructureUtil.stateBuilderAltair().slot(UInt64.ZERO).build();
+    assertThat(
+            syncCommitteeUtil.isStateUsableForCommitteeCalculationAtEpoch(
+                state, UInt64.valueOf(epochsPerPeriod).times(2)))
+        .isFalse();
+  }
+
+  @Test
+  void isStateUsableForCommitteeCalculationAtEpoch_shouldBeTrueWhenStateFromRequestedSyncPeriod() {
+    final BeaconStateAltair state =
+        dataStructureUtil.stateBuilderAltair().slot(UInt64.ZERO).build();
+    assertThat(syncCommitteeUtil.isStateUsableForCommitteeCalculationAtEpoch(state, UInt64.ONE))
+        .isTrue();
+  }
+
+  @Test
+  void isStateUsableForCommitteeCalculationAtEpoch_shouldBeTrueWhenStateFromPreviousSyncPeriod() {
+    final int epochsPerPeriod = config.getEpochsPerSyncCommitteePeriod();
+
+    final BeaconStateAltair state =
+        dataStructureUtil.stateBuilderAltair().slot(UInt64.ZERO).build();
+    assertThat(
+            syncCommitteeUtil.isStateUsableForCommitteeCalculationAtEpoch(
+                state, UInt64.valueOf(epochsPerPeriod).times(1)))
+        .isTrue();
   }
 
   private BeaconState createStateWithCurrentSyncCommittee(

--- a/validator/client/src/main/java/tech/pegasys/teku/validator/client/SyncCommitteeScheduler.java
+++ b/validator/client/src/main/java/tech/pegasys/teku/validator/client/SyncCommitteeScheduler.java
@@ -166,11 +166,16 @@ public class SyncCommitteeScheduler implements ValidatorTimingChannel {
     }
 
     public void calculateDuties() {
+
+      // Always use the last epoch in the period since it's the most likely to still be in-memory
+      // This also handles the case where the fork slot is within the sync committee period by
+      // ensuring that we request an epoch after the fork slot has occurred.
       duties =
           duties.or(
               () ->
                   Optional.of(
-                      PendingDuties.calculateDuties(metricsSystem, dutyLoader, periodStartEpoch)));
+                      PendingDuties.calculateDuties(
+                          metricsSystem, dutyLoader, nextPeriodStartEpoch.minusMinZero(1))));
     }
 
     public void recalculate() {

--- a/validator/client/src/test/java/tech/pegasys/teku/validator/client/SyncCommitteeSchedulerTest.java
+++ b/validator/client/src/test/java/tech/pegasys/teku/validator/client/SyncCommitteeSchedulerTest.java
@@ -71,23 +71,24 @@ class SyncCommitteeSchedulerTest {
   void shouldCalculateCurrentPeriodDutiesOnFirstSlot() {
     scheduler.onSlot(UInt64.ONE);
 
-    verify(dutyLoader).loadDutiesForEpoch(UInt64.ZERO);
+    // Should request the last epoch in the period as it's most likely to be non-finalized
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(0));
   }
 
   @Test
   void shouldNotReloadDutiesAtNextEpochWhenInSameSyncCommitteePeriod() {
     scheduler.onSlot(UInt64.ONE);
-    verify(dutyLoader).loadDutiesForEpoch(UInt64.ZERO);
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(0));
 
     scheduler.onSlot(spec.computeStartSlotAtEpoch(UInt64.ONE));
-    verify(dutyLoader, never()).loadDutiesForEpoch(UInt64.ONE);
+    verifyNoMoreInteractions(dutyLoader);
   }
 
   @Test
   void shouldPerformProductionForEachSlotWhenAttestationCreationDue() {
     scheduler.onSlot(UInt64.ONE);
 
-    requestedDutiesByEpoch.get(UInt64.ZERO).complete(Optional.of(duties));
+    getRequestedDutiesForSyncCommitteePeriod(0).complete(Optional.of(duties));
 
     UInt64.range(UInt64.ONE, UInt64.valueOf(10))
         .forEach(
@@ -101,7 +102,7 @@ class SyncCommitteeSchedulerTest {
   void shouldPerformAggregationForEachSlotWhenAttestationAggregationDue() {
     scheduler.onSlot(UInt64.ONE);
 
-    requestedDutiesByEpoch.get(UInt64.ZERO).complete(Optional.of(duties));
+    getRequestedDutiesForSyncCommitteePeriod(0).complete(Optional.of(duties));
 
     UInt64.range(UInt64.ONE, UInt64.valueOf(10))
         .forEach(
@@ -119,10 +120,10 @@ class SyncCommitteeSchedulerTest {
     final UInt64 subscribeEpoch = nextSyncCommitteePeriodStartEpoch.minus(4);
 
     scheduler.onSlot(UInt64.ONE);
-    verify(dutyLoader, never()).loadDutiesForEpoch(nextSyncCommitteePeriodStartEpoch);
+    verify(dutyLoader, never()).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(1));
     scheduler.onSlot(spec.computeStartSlotAtEpoch(subscribeEpoch));
 
-    verify(dutyLoader).loadDutiesForEpoch(nextSyncCommitteePeriodStartEpoch);
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(1));
   }
 
   @Test
@@ -146,10 +147,10 @@ class SyncCommitteeSchedulerTest {
     final UInt64 subscribeSlot = spec.computeStartSlotAtEpoch(subscribeEpoch);
 
     scheduler.onSlot(UInt64.ONE);
-    verify(dutyLoader).loadDutiesForEpoch(UInt64.ZERO);
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(0));
 
     scheduler.onSlot(subscribeSlot);
-    verify(dutyLoader).loadDutiesForEpoch(nextSyncCommitteePeriodStartEpoch);
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(1));
     verifyNoMoreInteractions(dutyLoader);
 
     // Already calculated all the duties we need so don't calculate them again
@@ -170,12 +171,12 @@ class SyncCommitteeSchedulerTest {
     final SyncCommitteeScheduledDuties nextDuties = createScheduledDuties();
 
     scheduler.onSlot(UInt64.valueOf(5));
-    verify(dutyLoader).loadDutiesForEpoch(UInt64.ZERO);
-    requestedDutiesByEpoch.get(UInt64.ZERO).complete(Optional.of(duties));
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(0));
+    getRequestedDutiesForSyncCommitteePeriod(0).complete(Optional.of(duties));
 
     scheduler.onSlot(subscribeSlot);
-    verify(dutyLoader).loadDutiesForEpoch(nextSyncCommitteePeriodStartEpoch);
-    requestedDutiesByEpoch.get(nextSyncCommitteePeriodStartEpoch).complete(Optional.of(nextDuties));
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(1));
+    getRequestedDutiesForSyncCommitteePeriod(1).complete(Optional.of(nextDuties));
 
     // Subscribed, but still performing duties for first sync committee period
     scheduler.onAttestationCreationDue(subscribeSlot);
@@ -204,8 +205,8 @@ class SyncCommitteeSchedulerTest {
 
     // Trigger calculation of duties for both current and next periods
     scheduler.onSlot(subscribeSlot);
-    requestedDutiesByEpoch.get(UInt64.ZERO).complete(Optional.of(duties));
-    requestedDutiesByEpoch.get(nextSyncCommitteePeriodStartEpoch).complete(Optional.of(nextDuties));
+    getRequestedDutiesForSyncCommitteePeriod(0).complete(Optional.of(duties));
+    getRequestedDutiesForSyncCommitteePeriod(1).complete(Optional.of(nextDuties));
 
     // Unexpectedly jump ahead to the next committee period without getting a slot event first
     scheduler.onAttestationCreationDue(nextSyncCommitteePeriodStartSlot);
@@ -232,12 +233,23 @@ class SyncCommitteeSchedulerTest {
   @Test
   void shouldRecalculateDutiesWhenBeaconNodeRestarts() {
     scheduler.onSlot(UInt64.ZERO);
-    verify(dutyLoader).loadDutiesForEpoch(UInt64.ZERO);
+    verify(dutyLoader).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(0));
 
     // Reconnecting the event stream may mean the node restarted so recalculate duties to ensure
     // subscriptions are refreshed
     scheduler.onPossibleMissedEvents();
-    verify(dutyLoader, times(2)).loadDutiesForEpoch(UInt64.ZERO);
+    verify(dutyLoader, times(2)).loadDutiesForEpoch(getRequestEpochForCommitteePeriod(0));
+  }
+
+  private SafeFuture<Optional<SyncCommitteeScheduledDuties>>
+      getRequestedDutiesForSyncCommitteePeriod(final int syncCommitteePeriod) {
+    final UInt64 requestEpoch = getRequestEpochForCommitteePeriod(syncCommitteePeriod);
+    return requestedDutiesByEpoch.get(requestEpoch);
+  }
+
+  private UInt64 getRequestEpochForCommitteePeriod(final int syncCommitteePeriod) {
+    return UInt64.valueOf(epochsPerSyncCommitteePeriod * (syncCommitteePeriod + 1L))
+        .minusMinZero(1);
   }
 
   private SyncCommitteeScheduledDuties createScheduledDuties() {

